// graphile-build-pg/node8plus/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "PgBasicsPlugin", {
  enumerable: true,
  get: function () {
    return _PgBasicsPlugin.default;
  }
});
Object.defineProperty(exports, "PgIntrospectionPlugin", {
  enumerable: true,
  get: function () {
    return _PgIntrospectionPlugin.default;
  }
});
Object.defineProperty(exports, "PgEntityKind", {
  enumerable: true,
  get: function () {
    return _PgIntrospectionPlugin.PgEntityKind;
  }
});
Object.defineProperty(exports, "PgTypesPlugin", {
  enumerable: true,
  get: function () {
    return _PgTypesPlugin.default;
  }
});
Object.defineProperty(exports, "PgJWTPlugin", {
  enumerable: true,
  get: function () {
    return _PgJWTPlugin.default;
  }
});
Object.defineProperty(exports, "PgTablesPlugin", {
  enumerable: true,
  get: function () {
    return _PgTablesPlugin.default;
  }
});
Object.defineProperty(exports, "PgConnectionArgFirstLastBeforeAfter", {
  enumerable: true,
  get: function () {
    return _PgConnectionArgFirstLastBeforeAfter.default;
  }
});
Object.defineProperty(exports, "PgConnectionArgOrderBy", {
  enumerable: true,
  get: function () {
    return _PgConnectionArgOrderBy.default;
  }
});
Object.defineProperty(exports, "PgConnectionArgOrderByDefaultValue", {
  enumerable: true,
  get: function () {
    return _PgConnectionArgOrderByDefaultValue.default;
  }
});
Object.defineProperty(exports, "PgConnectionArgCondition", {
  enumerable: true,
  get: function () {
    return _PgConnectionArgCondition.default;
  }
});
Object.defineProperty(exports, "PgConditionComputedColumnPlugin", {
  enumerable: true,
  get: function () {
    return _PgConditionComputedColumnPlugin.default;
  }
});
Object.defineProperty(exports, "PgAllRows", {
  enumerable: true,
  get: function () {
    return _PgAllRows.default;
  }
});
Object.defineProperty(exports, "PgColumnsPlugin", {
  enumerable: true,
  get: function () {
    return _PgColumnsPlugin.default;
  }
});
Object.defineProperty(exports, "PgColumnDeprecationPlugin", {
  enumerable: true,
  get: function () {
    return _PgColumnDeprecationPlugin.default;
  }
});
Object.defineProperty(exports, "PgForwardRelationPlugin", {
  enumerable: true,
  get: function () {
    return _PgForwardRelationPlugin.default;
  }
});
Object.defineProperty(exports, "PgBackwardRelationPlugin", {
  enumerable: true,
  get: function () {
    return _PgBackwardRelationPlugin.default;
  }
});
Object.defineProperty(exports, "PgRowByUniqueConstraint", {
  enumerable: true,
  get: function () {
    return _PgRowByUniqueConstraint.default;
  }
});
Object.defineProperty(exports, "PgComputedColumnsPlugin", {
  enumerable: true,
  get: function () {
    return _PgComputedColumnsPlugin.default;
  }
});
Object.defineProperty(exports, "PgQueryProceduresPlugin", {
  enumerable: true,
  get: function () {
    return _PgQueryProceduresPlugin.default;
  }
});
Object.defineProperty(exports, "PgOrderAllColumnsPlugin", {
  enumerable: true,
  get: function () {
    return _PgOrderAllColumnsPlugin.default;
  }
});
Object.defineProperty(exports, "PgOrderComputedColumnsPlugin", {
  enumerable: true,
  get: function () {
    return _PgOrderComputedColumnsPlugin.default;
  }
});
Object.defineProperty(exports, "PgOrderByPrimaryKeyPlugin", {
  enumerable: true,
  get: function () {
    return _PgOrderByPrimaryKeyPlugin.default;
  }
});
Object.defineProperty(exports, "PgRowNode", {
  enumerable: true,
  get: function () {
    return _PgRowNode.default;
  }
});
Object.defineProperty(exports, "PgNodeAliasPostGraphile", {
  enumerable: true,
  get: function () {
    return _PgNodeAliasPostGraphile.default;
  }
});
Object.defineProperty(exports, "PgNodeAliasPostGraphQL", {
  enumerable: true,
  get: function () {
    return _PgNodeAliasPostGraphile.default;
  }
});
Object.defineProperty(exports, "PgRecordReturnTypesPlugin", {
  enumerable: true,
  get: function () {
    return _PgRecordReturnTypesPlugin.default;
  }
});
Object.defineProperty(exports, "PgRecordFunctionConnectionPlugin", {
  enumerable: true,
  get: function () {
    return _PgRecordFunctionConnectionPlugin.default;
  }
});
Object.defineProperty(exports, "PgScalarFunctionConnectionPlugin", {
  enumerable: true,
  get: function () {
    return _PgScalarFunctionConnectionPlugin.default;
  }
});
Object.defineProperty(exports, "PageInfoStartEndCursor", {
  enumerable: true,
  get: function () {
    return _PageInfoStartEndCursor.default;
  }
});
Object.defineProperty(exports, "PgConnectionTotalCount", {
  enumerable: true,
  get: function () {
    return _PgConnectionTotalCount.default;
  }
});
Object.defineProperty(exports, "PgMutationCreatePlugin", {
  enumerable: true,
  get: function () {
    return _PgMutationCreatePlugin.default;
  }
});
Object.defineProperty(exports, "PgMutationUpdateDeletePlugin", {
  enumerable: true,
  get: function () {
    return _PgMutationUpdateDeletePlugin.default;
  }
});
Object.defineProperty(exports, "PgMutationProceduresPlugin", {
  enumerable: true,
  get: function () {
    return _PgMutationProceduresPlugin.default;
  }
});
Object.defineProperty(exports, "PgMutationPayloadEdgePlugin", {
  enumerable: true,
  get: function () {
    return _PgMutationPayloadEdgePlugin.default;
  }
});
Object.defineProperty(exports, "parseIdentifier", {
  enumerable: true,
  get: function () {
    return _parseIdentifier.default;
  }
});
Object.defineProperty(exports, "omit", {
  enumerable: true,
  get: function () {
    return _omit.default;
  }
});
Object.defineProperty(exports, "formatSQLForDebugging", {
  enumerable: true,
  get: function () {
    return _debugSql.formatSQLForDebugging;
  }
});
Object.defineProperty(exports, "upperFirst", {
  enumerable: true,
  get: function () {
    return _graphileBuild.upperFirst;
  }
});
Object.defineProperty(exports, "camelCase", {
  enumerable: true,
  get: function () {
    return _graphileBuild.camelCase;
  }
});
Object.defineProperty(exports, "constantCase", {
  enumerable: true,
  get: function () {
    return _graphileBuild.constantCase;
  }
});
Object.defineProperty(exports, "parseTags", {
  enumerable: true,
  get: function () {
    return _utils.parseTags;
  }
});
Object.defineProperty(exports, "withPgClient", {
  enumerable: true,
  get: function () {
    return _withPgClient.default;
  }
});
Object.defineProperty(exports, "getPgClientAndReleaserFromConfig", {
  enumerable: true,
  get: function () {
    return _withPgClient.getPgClientAndReleaserFromConfig;
  }
});
exports.inflections = exports.defaultPlugins = void 0;

var _PgBasicsPlugin = _interopRequireDefault(require("./plugins/PgBasicsPlugin"));

var _PgIntrospectionPlugin = _interopRequireWildcard(require("./plugins/PgIntrospectionPlugin"));

var _PgTypesPlugin = _interopRequireDefault(require("./plugins/PgTypesPlugin"));

var _PgJWTPlugin = _interopRequireDefault(require("./plugins/PgJWTPlugin"));

var _PgTablesPlugin = _interopRequireDefault(require("./plugins/PgTablesPlugin"));

var _PgConnectionArgFirstLastBeforeAfter = _interopRequireDefault(require("./plugins/PgConnectionArgFirstLastBeforeAfter"));

var _PgConnectionArgOrderBy = _interopRequireDefault(require("./plugins/PgConnectionArgOrderBy"));

var _PgConnectionArgOrderByDefaultValue = _interopRequireDefault(require("./plugins/PgConnectionArgOrderByDefaultValue"));

var _PgConnectionArgCondition = _interopRequireDefault(require("./plugins/PgConnectionArgCondition"));

var _PgConditionComputedColumnPlugin = _interopRequireDefault(require("./plugins/PgConditionComputedColumnPlugin"));

var _PgAllRows = _interopRequireDefault(require("./plugins/PgAllRows"));

var _PgColumnsPlugin = _interopRequireDefault(require("./plugins/PgColumnsPlugin"));

var _PgColumnDeprecationPlugin = _interopRequireDefault(require("./plugins/PgColumnDeprecationPlugin"));

var _PgForwardRelationPlugin = _interopRequireDefault(require("./plugins/PgForwardRelationPlugin"));

var _PgBackwardRelationPlugin = _interopRequireDefault(require("./plugins/PgBackwardRelationPlugin"));

var _PgRowByUniqueConstraint = _interopRequireDefault(require("./plugins/PgRowByUniqueConstraint"));

var _PgComputedColumnsPlugin = _interopRequireDefault(require("./plugins/PgComputedColumnsPlugin"));

var _PgQueryProceduresPlugin = _interopRequireDefault(require("./plugins/PgQueryProceduresPlugin"));

var _PgOrderAllColumnsPlugin = _interopRequireDefault(require("./plugins/PgOrderAllColumnsPlugin"));

var _PgOrderComputedColumnsPlugin = _interopRequireDefault(require("./plugins/PgOrderComputedColumnsPlugin"));

var _PgOrderByPrimaryKeyPlugin = _interopRequireDefault(require("./plugins/PgOrderByPrimaryKeyPlugin"));

var _PgRowNode = _interopRequireDefault(require("./plugins/PgRowNode"));

var _PgNodeAliasPostGraphile = _interopRequireDefault(require("./plugins/PgNodeAliasPostGraphile"));

var _PgRecordReturnTypesPlugin = _interopRequireDefault(require("./plugins/PgRecordReturnTypesPlugin"));

var _PgRecordFunctionConnectionPlugin = _interopRequireDefault(require("./plugins/PgRecordFunctionConnectionPlugin"));

var _PgScalarFunctionConnectionPlugin = _interopRequireDefault(require("./plugins/PgScalarFunctionConnectionPlugin"));

var _PageInfoStartEndCursor = _interopRequireDefault(require("./plugins/PageInfoStartEndCursor"));

var _PgConnectionTotalCount = _interopRequireDefault(require("./plugins/PgConnectionTotalCount"));

var _PgMutationCreatePlugin = _interopRequireDefault(require("./plugins/PgMutationCreatePlugin"));

var _PgMutationUpdateDeletePlugin = _interopRequireDefault(require("./plugins/PgMutationUpdateDeletePlugin"));

var _PgMutationProceduresPlugin = _interopRequireDefault(require("./plugins/PgMutationProceduresPlugin"));

var _PgMutationPayloadEdgePlugin = _interopRequireDefault(require("./plugins/PgMutationPayloadEdgePlugin"));

var inflections = _interopRequireWildcard(require("./inflections"));

exports.inflections = inflections;

var _parseIdentifier = _interopRequireDefault(require("./parseIdentifier"));

var _omit = _interopRequireDefault(require("./omit"));

var _debugSql = require("./plugins/debugSql");

var _graphileBuild = require("graphile-build");

var _utils = require("./utils");

var _withPgClient = _interopRequireWildcard(require("./withPgClient"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function () { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Mutations
const defaultPlugins = [_PgBasicsPlugin.default, _PgIntrospectionPlugin.default, _PgTypesPlugin.default, _PgJWTPlugin.default, _PgTablesPlugin.default, _PgConnectionArgFirstLastBeforeAfter.default, _PgConnectionArgOrderBy.default, _PgConnectionArgOrderByDefaultValue.default, _PgConnectionArgCondition.default, _PgConditionComputedColumnPlugin.default, _PgAllRows.default, _PgColumnsPlugin.default, _PgColumnDeprecationPlugin.default, _PgForwardRelationPlugin.default, _PgBackwardRelationPlugin.default, _PgRowByUniqueConstraint.default, _PgComputedColumnsPlugin.default, _PgQueryProceduresPlugin.default, _PgOrderAllColumnsPlugin.default, _PgOrderComputedColumnsPlugin.default, _PgOrderByPrimaryKeyPlugin.default, _PgRowNode.default, _PgNodeAliasPostGraphile.default, _PgRecordReturnTypesPlugin.default, _PgRecordFunctionConnectionPlugin.default, _PgScalarFunctionConnectionPlugin.default, // For PostGraphile compatibility
_PageInfoStartEndCursor.default, // For PostGraphile compatibility
_PgConnectionTotalCount.default, // Mutations
_PgMutationCreatePlugin.default, _PgMutationUpdateDeletePlugin.default, _PgMutationProceduresPlugin.default, _PgMutationPayloadEdgePlugin.default];
exports.defaultPlugins = defaultPlugins;
