// graphile-utils/node8plus/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.orderByAscDesc = exports.makeAddPgTableOrderByPlugin = exports.makeAddPgTableConditionPlugin = exports.makeProcessSchemaPlugin = exports.makeChangeNullabilityPlugin = exports.makeWrapResolversPlugin = exports.makePluginByCombiningPlugins = exports.makeExtendSchemaPlugin = exports.makeAddInflectorsPlugin = exports.gql = exports.embed = void 0;
const tslib_1 = require("tslib");
const gql_1 = require("./gql");
Object.defineProperty(exports, "embed", { enumerable: true, get: function () { return gql_1.embed; } });
Object.defineProperty(exports, "gql", { enumerable: true, get: function () { return gql_1.gql; } });
const makeAddInflectorsPlugin_1 = require("./makeAddInflectorsPlugin");
exports.makeAddInflectorsPlugin = makeAddInflectorsPlugin_1.default;
const makeExtendSchemaPlugin_1 = require("./makeExtendSchemaPlugin");
exports.makeExtendSchemaPlugin = makeExtendSchemaPlugin_1.default;
const makePluginByCombiningPlugins_1 = require("./makePluginByCombiningPlugins");
exports.makePluginByCombiningPlugins = makePluginByCombiningPlugins_1.default;
const makeWrapResolversPlugin_1 = require("./makeWrapResolversPlugin");
exports.makeWrapResolversPlugin = makeWrapResolversPlugin_1.default;
const makeChangeNullabilityPlugin_1 = require("./makeChangeNullabilityPlugin");
exports.makeChangeNullabilityPlugin = makeChangeNullabilityPlugin_1.default;
const makeProcessSchemaPlugin_1 = require("./makeProcessSchemaPlugin");
exports.makeProcessSchemaPlugin = makeProcessSchemaPlugin_1.default;
const makeAddPgTableConditionPlugin_1 = require("./makeAddPgTableConditionPlugin");
exports.makeAddPgTableConditionPlugin = makeAddPgTableConditionPlugin_1.default;
const makeAddPgTableOrderByPlugin_1 = require("./makeAddPgTableOrderByPlugin");
exports.makeAddPgTableOrderByPlugin = makeAddPgTableOrderByPlugin_1.default;
Object.defineProperty(exports, "orderByAscDesc", { enumerable: true, get: function () { return makeAddPgTableOrderByPlugin_1.orderByAscDesc; } });
tslib_1.__exportStar(require("./makePgSmartTagsPlugin"), exports);
